src/Swarm/Game/Location.hs

```diff
@@ -71,11 +90,68 @@ deriving instance FromJSON (V2 Int32)
 deriving instance FromJSONKey (V2 Int32)
 deriving instance ToJSONKey (V2 Int32)
 
-instance FromJSON Location where
-  parseJSON = fmap P . parseJSON
-
-instance ToJSON Location where
-  toJSON (P v) = toJSON v
+toHeading :: AbsoluteDir -> Heading
+toHeading = \case
+  DNorth -> north
+  DSouth -> south
+  DEast -> east
+  DWest -> west
+
+-- | The cardinal direction north = @V2 0 1@.
+north :: Heading
+north = V2 0 1
+
+-- | The cardinal direction south = @V2 0 (-1)@.
+south :: Heading
+south = V2 0 (-1)
+
+-- | The cardinal direction east = @V2 1 0@.
+east :: Heading
+east = V2 1 0
+
+-- | The cardinal direction west = @V2 (-1) 0@.
+west :: Heading
+west = V2 (-1) 0
+
+-- | The direction for viewing the current cell = @V2 0 0@.
+down :: Heading
+down = zero
+
+-- | The 'applyTurn' function gives the meaning of each 'Direction' by
+--   turning relative to the given heading or by turning to an absolute
+--   heading
+applyTurn :: Direction -> Heading -> Heading
+applyTurn d = case d of
+  DRelative e -> case e of
+    DLeft -> relative perp
+    DRight -> relative (fmap negate . perp)
```

Comment:
I've been meaning to replace `fmap negate` with just [`negated`](https://hackage.haskell.org/package/linear-1.22/docs/Linear-Vector.html#v:negated)

Reply:
Neat, thanks! :+1: